tests/paddle/test_operators.py

```diff
@@ -739,31 +741,49 @@ def _get_fused_attention_out(self):
 
         return out, q_grad, k_grad, v_grad
 
-    @pytest.mark.skipif(paddle.device.cuda.get_device_capability() not in ((8, 0), (9, 0)),
-                        reason="cuDNN fMHA requires Ampere and Hopper GPU")
     @pytest.mark.parametrize('b, s, h, d', SELF_ATTN_CASES)
     @pytest.mark.parametrize('dtype', ['float16', 'bfloat16'])
     @pytest.mark.parametrize('is_causal_masking', [True, False])
     def test_self_attn_forward_backward(self, b, s, h, d, dtype, is_causal_masking):
         """
         test self attention forward + backward
         """
+        if not is_fused_attention_supported(
+            head_size=d,
+            q_seqlen=s,
+            kv_seqlen=s,
+            dtype=dtype,
+            dropout=0.0,
+            qkv_layout="qkv_interleaved",
+            bias_type="no_bias",
+            mask_type="causal" if is_causal_masking else "padding",
+        ):
+            pytest.skip("cuDNN fused attention is not supported")
         self.set_input(b, s, s, h, d, dtype, "self_attn", is_causal_masking)
         reference_out, q_grad_ref, k_grad_ref, v_grad_ref = self._get_reference_out()
         fused_attention_out, q_grad, k_grad, v_grad = self._get_fused_attention_out()
-        assert_allclose(reference_out, fused_attention_out, rtol=1e-3, atol=1e-2)
-        assert_allclose(q_grad_ref, q_grad, rtol=1e-3, atol=1e-2)
-        assert_allclose(k_grad_ref, k_grad, rtol=1e-3, atol=1e-2)
-        assert_allclose(v_grad_ref, v_grad, rtol=1e-3, atol=1e-2)
+        assert_allclose(reference_out, fused_attention_out, rtol=2.5e-2, atol=2.5e-2)
+        assert_allclose(q_grad_ref, q_grad, rtol=2.5e-2, atol=2.5e-2)
+        assert_allclose(k_grad_ref, k_grad, rtol=2.5e-2, atol=2.5e-2)
+        assert_allclose(v_grad_ref, v_grad, rtol=2.5e-2, atol=2.5e-2)
```

Comment:
These tolerances match the PyTorch tests:
https://github.com/NVIDIA/TransformerEngine/blob/daa5e1840571be4f9cd11053bbde3f1bfd8ca489/tests/pytorch/test_fused_attn.py#L86

I suspect there might be a bug in how cuDNN fused attention handles causal masks. With the old tolerances, the test passes on H100 (max-512 seqlen backend) and fails on L40 (arbitrary backend). I tried changing the attention mask so that it is a plain triangular matrix that does not take the actual sequence lengths into account, i.e. modified 
https://github.com/NVIDIA/TransformerEngine/blob/777e45be0b36206557c76280451bf94942637115/tests/paddle/test_operators.py#L605
to loop over `range(self.q_seqlen)`. This resulted in the test passing on L40 and failing on H100. This makes me think that the two backends are subtly inconsistent. I'll continue digging in, but if the fix is difficult then I think we should consolidate these changes and incorporate the fix in a future PR.

Pinging @cyanguwa.

Reply:
It turns out this is because the artibrary seqlen backend assumes that all seqlens are equal. However, this inconsistency is handled in https://github.com/NVIDIA/TransformerEngine/pull/403, which has more expressive layout configurations. Attempting to run with a variable seqlen format (e.g. `TH3D`) on a system that only supports the arbitrary seqlen backend will result in a runtime error. Upgrading Paddle to use the new attention layout formats is a higher priority than fixing this edge case.